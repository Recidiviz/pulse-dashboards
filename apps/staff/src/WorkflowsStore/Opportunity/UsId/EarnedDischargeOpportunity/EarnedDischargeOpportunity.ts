// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { DocumentData } from "firebase/firestore";

import { OpportunityUpdateWithForm } from "../../../../FirestoreStore";
import { Client } from "../../../Client";
import { UsIdEarnedDischargeForm } from "../../Forms/UsIdEarnedDischargeForm";
import { OpportunityBase } from "../../OpportunityBase";
import {
  EarnedDischargeDraftData,
  EarnedDischargeReferralRecord,
  usIdEarnedDischargeSchema,
} from "./EarnedDischargeReferralRecord";

export class EarnedDischargeOpportunity extends OpportunityBase<
  Client,
  EarnedDischargeReferralRecord,
  OpportunityUpdateWithForm<EarnedDischargeDraftData>
> {
  form?: UsIdEarnedDischargeForm;

  constructor(client: Client, record: DocumentData) {
    super(
      client,
      "earnedDischarge",
      client.rootStore,
      usIdEarnedDischargeSchema.parse(record),
    );

    this.form = new UsIdEarnedDischargeForm(this, client.rootStore);
  }

  get almostEligibleStatusMessage(): string | undefined {
    const { requirementsAlmostMet } = this;
    if (requirementsAlmostMet.length === 0) return;
    return requirementsAlmostMet[0].text;
  }

  get eligibilityDate(): Date | undefined {
    if (!this.record) return;
    return super.eligibilityDate ?? this.record.eligibleStartDate;
  }
}
