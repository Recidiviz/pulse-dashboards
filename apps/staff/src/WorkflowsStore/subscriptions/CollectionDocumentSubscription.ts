// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { doc, DocumentData, DocumentReference } from "firebase/firestore";

import FirestoreStore from "../../FirestoreStore";
import { FIRESTORE_COLLECTIONS_MAP } from "../../FirestoreStore/constants";
import type { FirestoreCollectionKey } from "../../FirestoreStore/types";
import { FirestoreDocumentSubscription } from "./FirestoreDocumentSubscription";
import { TransformFunction, UpdateFunction, ValidateFunction } from "./types";

/**
 * Subscribes to the specified document in the specified collection.
 * RecordType is essentially just an assertion for downstream convenience;
 * the data source itself is not validated or even strongly typed. Proceed with caution!
 */
export class CollectionDocumentSubscription<
  RecordType extends DocumentData
> extends FirestoreDocumentSubscription<RecordType> {
  readonly dataSource: DocumentReference;

  constructor(
    firestoreStore: FirestoreStore,
    firestoreCollectionKey: FirestoreCollectionKey,
    recordId: string,
    transformFn?: TransformFunction<RecordType>,
    validateFn?: ValidateFunction<RecordType>,
    updateFn?: UpdateFunction<DocumentData>
  ) {
    super(transformFn, validateFn, updateFn);

    this.dataSource = doc(
      firestoreStore.db,
      FIRESTORE_COLLECTIONS_MAP[firestoreCollectionKey],
      recordId
    );
  }
}
