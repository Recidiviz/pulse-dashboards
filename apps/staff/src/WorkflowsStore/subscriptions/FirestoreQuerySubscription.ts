// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import * as Sentry from "@sentry/react";
import {
  DocumentData,
  onSnapshot,
  Query,
  QuerySnapshot,
  Unsubscribe,
} from "firebase/firestore";
import {
  action,
  computed,
  makeObservable,
  override,
  reaction,
  runInAction,
} from "mobx";
import { IDisposer } from "mobx-utils";

import { FeatureGateError } from "../../errors";
import { FirestoreSubscription } from "./FirestoreSubscription";
import {
  QuerySubscription,
  TransformFunction,
  ValidateFunction,
} from "./types";
import { defaultTransformFunction, defaultValidateFunction } from "./utils";

export abstract class FirestoreQuerySubscription<
    DataFormat extends DocumentData,
  >
  extends FirestoreSubscription<DataFormat[]>
  implements QuerySubscription<DataFormat>
{
  abstract get dataSource(): Query | undefined;

  protected disposeDynamicDataSource?: IDisposer;

  constructor(
    protected transformRecord: TransformFunction<DataFormat> = defaultTransformFunction,
    protected validateRecord: ValidateFunction<DataFormat> = defaultValidateFunction,
  ) {
    super([]);

    makeObservable<this, "updateData">(this, {
      dataSource: computed,
      updateData: action,
      subscribe: override,
      unsubscribe: override,
    });
  }

  protected updateData(snapshot: QuerySnapshot | undefined): void {
    const docs: DataFormat[] = [];

    // receiving no snapshot at all is a special case that indicates there is no valid data source;
    // when this happens we want to reset the hydration state in addition to clearing out any data
    // that may already exist from previous updates
    if (!snapshot) {
      this.data = docs;
      this.hydrationState = { status: "needs hydration" };
      return;
    }

    const errors: unknown[] = [];

    snapshot.forEach((doc) => {
      try {
        const record = this.transformRecord(
          doc.data({ serverTimestamps: "estimate" }),
        );

        if (record !== undefined) this.validateRecord(record);

        if (record !== undefined) {
          docs.push(record);
        }
      } catch (e) {
        // Move on to allow for partial success
        errors.push(e);
        // don't log routine feature flag checks, but do log everything else
        if (!(e instanceof FeatureGateError)) {
          Sentry.captureException(e);
        }
      }
    });

    this.data = docs;
    // regardless of errors we always consider this hydrated
    this.hydrationState = { status: "hydrated" };
  }

  protected startSnapshotListener(): Unsubscribe | undefined {
    if (!this.dataSource) {
      this.updateData(undefined);
    } else {
      return onSnapshot(
        this.dataSource,
        (result) => this.updateData(result),
        (error) => {
          runInAction(() => {
            // an error that occurs after data is fetched is considered a partial failure,
            // but this one means that the fetch operation itself failed and therefore so does hydration.
            this.hydrationState = { status: "failed", error };
            this.data = [];
          });
          Sentry.captureException(error);
        },
      );
    }
  }

  subscribe(): void {
    super.subscribe();

    // watches dataSource for changes and re-subscribes accordingly
    this.disposeDynamicDataSource = reaction(
      () => this.dataSource,
      () => {
        this.unsubscribe();
        this.subscribe();
      },
    );
  }

  unsubscribe(): void {
    super.unsubscribe();
    this.disposeDynamicDataSource?.();
  }
}
