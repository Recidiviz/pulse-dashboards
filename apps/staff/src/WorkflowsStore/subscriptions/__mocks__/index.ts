// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { MockDocumentSubscription } from "./MockDocumentSubscription";
import { MockQuerySubscription } from "./MockQuerySubscription";

export const CollectionDocumentSubscription = MockDocumentSubscription;

export const OpportunityUpdateSubscription = MockDocumentSubscription;

export const CaseloadSubscription = MockQuerySubscription;

export const StaffSubscription = MockQuerySubscription;

export const UserSubscription = MockQuerySubscription;

export const LocationSubscription = MockQuerySubscription;
