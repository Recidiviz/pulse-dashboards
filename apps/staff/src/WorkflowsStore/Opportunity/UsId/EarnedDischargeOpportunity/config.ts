// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { OpportunityConfig } from "../../OpportunityConfigs";
import { EarnedDischargeOpportunity } from "./EarnedDischargeOpportunity";

export const usIdEarnedDischargeConfig: OpportunityConfig<EarnedDischargeOpportunity> =
  {
    systemType: "SUPERVISION",
    stateCode: "US_ID",
    urlSection: "earnedDischarge",
    label: "Earned Discharge",
    dynamicEligibilityText: "client[|s] may be eligible for earned discharge",
    callToAction: `Review clients who may be eligible for Earned Discharge and complete a pre-filled request form.`,
    subheading:
      "This alert helps staff identify people who may be eligible for earned discharge and directs staff to complete the pre-filled request form.",
    firestoreCollection: "US_ID-earnedDischargeReferrals",
    snooze: {
      defaultSnoozeDays: 30,
      maxSnoozeDays: 90,
    },
    methodologyUrl:
      "http://forms.idoc.idaho.gov/WebLink/0/edoc/282369/Termination%20of%20Probation%20or%20Parole%20Supervision.pdf",
    denialReasons: {
      SCNC: "Not compliant with special conditions",
      FFR: "Failure to make payments towards fines, fees, and restitution despite ability to pay",
      INTERLOCK: "Has an active interlock device",
      NCIC: "Did not pass NCIC check",
      PCD: "Parole Commission permanently denied early discharge request",
      CD: "Court permanently denied early discharge request",
      MIS: "Has had a violent misdemeanor conviction in the past 12 months",
      Other: "Other, please specify a reason",
    },
    sidebarComponents: ["ClientProfileDetails", "CaseNotes"],
    tooltipEligibilityText: "Eligible for Earned Discharge",
    eligibleCriteriaCopy: {
      usIdLsirLevelLowModerateForXDays: {
        text: '{{#if (eq "LOW" riskLevel)}}Currently low risk with no increase in risk level in past 90 days{{else}}Currently moderate risk with no increase in risk level in past 360 days{{/if}}',
        tooltip:
          "Assessed at low risk level on LSI-R with no risk increase in past 90 days or moderate risk level on LSI-R with no risk increase in past 360 days",
      },
      negativeDaWithin90Days: {
        text: "Negative UA within past 90 days",
        tooltip:
          "Negative UA within past 90 days, unless the client lacks a history of drug/alcohol abuse or has been supervised at low risk for more than one year",
      },
      usIdIncomeVerifiedWithin3Months: {
        text: "Verified compliant employment",
        tooltip:
          "Verified employment status, full-time student, or adequate lawful income from non-employment sources have been confirmed within past 3 months",
      },
      noFelonyWithin24Months: {
        text: "No felony convictions in past 24 months",
        tooltip:
          "Has not committed a felony while on probation or parole in past 24 months",
      },
      noViolentMisdemeanorWithin12Months: {
        text: "No violent misdemeanor convictions in past 12 months",
        tooltip:
          "Has not committed a violent misdemeanor or DUI misdemeanor while on probation or parole in past 12 months",
      },
      pastEarnedDischargeEligibleDate: {
        text: "Served {{daysToYearsMonthsPast (daysPast opportunity.person.supervisionStartDate)}}",
        tooltip:
          "If on probation, served minimum sentence according to the court; if on parole for a nonviolent crime, served at least one year; if on parole for a sex/violent offense, served at least one-third of remaining sentence; if on parole for a life sentence, served at least five years on parole",
      },
    },
    ineligibleCriteriaCopy: {
      pastEarnedDischargeEligibleDate: {
        text: "Needs {{monthsOrDaysRemainingFromToday eligibleDate}} on supervision",
        tooltip:
          "If on probation, served minimum sentence according to the court; if on parole for a nonviolent crime, served at least one year; if on parole for a sex/violent offense, served at least one-third of remaining sentence; if on parole for a life sentence, served at least five years on parole",
      },
    },
    homepagePosition: 2,
  };
