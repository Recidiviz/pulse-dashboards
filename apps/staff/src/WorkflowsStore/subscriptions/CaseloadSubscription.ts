// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  DocumentData,
  FieldPath,
  Query,
  query,
  where,
} from "firebase/firestore";
import { z } from "zod";

import { FirestoreCollectionKey } from "~firestore-api";

import { WorkflowsStore } from "../WorkflowsStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class CaseloadSubscription<
  RecordType extends DocumentData,
> extends FirestoreQuerySubscription<RecordType> {
  workflowsStore: WorkflowsStore;

  firestoreCollectionKey: FirestoreCollectionKey;

  personType: RecordType["personType"];

  constructor(
    workflowsStore: WorkflowsStore,
    firestoreCollectionKey: FirestoreCollectionKey,
    personType: RecordType["personType"],
    parser?: z.ZodType<RecordType, any, any>,
  ) {
    parser ? super(parser.parse) : super();
    this.workflowsStore = workflowsStore;
    this.firestoreCollectionKey = firestoreCollectionKey;
    this.personType = personType;
  }

  get dataSource(): Query | undefined {
    const {
      selectedSearchIds,
      rootStore: { currentTenantId, firestoreStore },
    } = this.workflowsStore;

    if (!currentTenantId || !selectedSearchIds.length) {
      return undefined;
    }

    const { firestoreCollectionKey, personType } = this;

    const config = this.workflowsStore.systemConfigFor(
      personType === "RESIDENT" ? "INCARCERATION" : "SUPERVISION",
    );

    const constraints = [
      where("stateCode", "==", currentTenantId),
      where(
        new FieldPath(...config.searchField),
        config.searchOp ?? "in",
        selectedSearchIds,
      ),
    ];

    if (config.onlySurfaceEligible) {
      constraints.push(where("allEligibleOpportunities", "!=", []));
    }

    return query(
      firestoreStore.collection(firestoreCollectionKey),
      ...constraints,
    ).withConverter({
      fromFirestore(snapshot, options) {
        const doc = snapshot.data(options);
        return { ...doc, recordId: snapshot.id, personType };
      },
      // these collections are read-only, so this should never be used, but it is required by Firestore
      toFirestore(record: any) {
        throw new Error(`Writing to ${firestoreCollectionKey} is not allowed`);
      },
    });
  }
}
