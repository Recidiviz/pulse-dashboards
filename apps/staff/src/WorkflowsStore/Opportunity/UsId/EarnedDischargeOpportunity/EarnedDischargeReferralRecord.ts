// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { camelCase, cloneDeep } from "lodash";

import { TransformFunction } from "../../../subscriptions";
import { fieldToDate } from "../../../utils";
import { WithCaseNotes } from "../../types";
import { transformCaseNotes } from "../../utils/caseNotesUtils";
import {
  LSUEarnedDischargeEligibleCriteria,
  LSUEarnedDischargeIneligibleCriteria,
  transformLSUEarnedDischargeCriteria,
} from "../LSUOpportunity/LSUReferralRecord";

export type EarnedDischargeReferralRecord = {
  stateCode: string;
  externalId: string;
  formInformation: {
    ncicCheckDate?: Date;
    fullTermReleaseDates?: Date[];
    chargeDescriptions?: string[];
    judgeNames?: {
      givenNames: string;
      middleNames: string;
      nameSuffix: string;
      surname: string;
    }[];
    countyNames?: string[];
    sentenceMax?: number[];
    sentenceMin?: number[];
    caseNumbers?: string[];
    dateImposed?: Date[];
    initialRestitution?: number;
    lastRestitutionPaymentDate?: Date;
    currentRestitutionBalance?: number;
    initialFines?: number;
    lastFinesPaymentDate?: Date;
    currentFinesBalance?: number;
    firstAssessmentScore?: number;
    firstAssessmentDate?: Date;
    latestAssessmentScore?: number;
    latestAssessmentDate?: Date;
  };
  ineligibleCriteria: Partial<LSUEarnedDischargeIneligibleCriteria>;
  eligibleCriteria: LSUEarnedDischargeEligibleCriteria & {
    usIdLsirLevelLowModerateForXDays: {
      eligibleDate: Date;
      riskLevel: "LOW" | "MODERATE";
    };
    pastEarnedDischargeEligibleDate?: {
      eligibleDate?: Date;
      sentenceType?: "PROBATION" | "PAROLE" | "DUAL";
    };
  };
  eligibleStartDate: Date;
} & WithCaseNotes;

export type EarnedDischargeCrimeTableKeys =
  | "judgeNames"
  | "countyNames"
  | "dateImposed"
  | "caseNumbers"
  | "chargeDescriptions"
  | "sentenceMin"
  | "sentenceMax"
  | "fullTermReleaseDates";

export type EarnedDischargeDraftData = {
  clientName: string;
  supervisionType: string;
  idocNumber: string;
  ftrDate: string;
  probationOfficerFullName: string;
  conditionCompliance: string;
  meetsIdocRequirements: string;
  ncicCheck: string;
  ncicCheckDate: string;
  initialRestitution: string;
  lastRestitutionPaymentDate: string;
  currentRestitutionBalance: string;
  initialFines: string;
  lastFinesPaymentDate: string;
  currentFinesBalance: string;
  firstAssessmentScore: string;
  firstAssessmentDate: string;
  latestAssessmentScore: string;
  latestAssessmentDate: string;
  numCrimeEntries: number;
} & Record<`${EarnedDischargeCrimeTableKeys}${number}`, string>;

export const transformEarnedDischargeReferral: TransformFunction<
  EarnedDischargeReferralRecord
> = (record) => {
  if (!record) {
    throw new Error("Record not found");
  }

  const transformedRecord = cloneDeep(record) as EarnedDischargeReferralRecord;
  const { eligibleCriteria, ineligibleCriteria } = record;

  const transformedCommonCriteria = transformLSUEarnedDischargeCriteria(record);

  transformedRecord.eligibleCriteria = {
    ...transformedRecord.eligibleCriteria,
    ...transformedCommonCriteria?.eligibleCriteria,
  };

  transformedRecord.ineligibleCriteria = {
    ...transformedRecord.ineligibleCriteria,
    ...transformedCommonCriteria?.ineligibleCriteria,
  };

  /* usIdLsirLevelLowModerateForXDays */
  transformedRecord.eligibleCriteria.usIdLsirLevelLowModerateForXDays = {
    riskLevel: eligibleCriteria.usIdLsirLevelLowModerateForXDays.riskLevel,
    eligibleDate: fieldToDate(
      eligibleCriteria.usIdLsirLevelLowModerateForXDays.eligibleDate,
    ),
  };

  /* Collapse `usIdParoleDualSupervisionPastEarlyDischargeDate` and `onProbationAtLeastOneYear` 
     into `pastEarnedDischargeEligibleDate`. */
  /* eligibleCriteria */
  if (
    eligibleCriteria?.usIdParoleDualSupervisionPastEarlyDischargeDate ||
    eligibleCriteria?.onProbationAtLeastOneYear
  ) {
    transformedRecord.eligibleCriteria.pastEarnedDischargeEligibleDate = {
      eligibleDate: fieldToDate(
        eligibleCriteria.usIdParoleDualSupervisionPastEarlyDischargeDate
          ?.eligibleDate ??
          eligibleCriteria.onProbationAtLeastOneYear?.eligibleDate,
      ),
      sentenceType:
        eligibleCriteria.usIdParoleDualSupervisionPastEarlyDischargeDate
          ?.sentenceType ??
        eligibleCriteria.onProbationAtLeastOneYear?.sentenceType,
    };
  }

  /* ineligibleCriteria */
  if (
    ineligibleCriteria?.usIdParoleDualSupervisionPastEarlyDischargeDate ||
    ineligibleCriteria?.onProbationAtLeastOneYear
  ) {
    transformedRecord.ineligibleCriteria.pastEarnedDischargeEligibleDate = {
      eligibleDate: fieldToDate(
        ineligibleCriteria.usIdParoleDualSupervisionPastEarlyDischargeDate
          ?.eligibleDate ??
          ineligibleCriteria.onProbationAtLeastOneYear?.eligibleDate,
      ),
      sentenceType:
        ineligibleCriteria.usIdParoleDualSupervisionPastEarlyDischargeDate
          ?.sentenceType ??
        ineligibleCriteria.onProbationAtLeastOneYear?.sentenceType,
    };
  }

  // delete vestigial criterion left over from TES we don't use in the front end
  delete (
    // @ts-expect-error
    // prettier-ignore
    transformedRecord.eligibleCriteria.usIdParoleDualSupervisionPastEarlyDischargeDate
  );
  delete (
    // @ts-expect-error
    // prettier-ignore
    transformedRecord.ineligibleCriteria.usIdParoleDualSupervisionPastEarlyDischargeDate
  );
  // @ts-expect-error
  delete transformedRecord.eligibleCriteria.onProbationAtLeastOneYear;
  // @ts-expect-error
  delete transformedRecord.ineligibleCriteria.onProbationAtLeastOneYear;

  delete (
    // @ts-expect-error
    transformedRecord.eligibleCriteria.supervisionNotPastFullTermCompletionDate
  );

  transformedRecord.eligibleStartDate = fieldToDate(record.eligibleStartDate);

  transformedRecord.caseNotes = transformCaseNotes(record.caseNotes);

  const {
    ncicCheckDate,
    dateImposed,
    fullTermReleaseDates,
    lastRestitutionPaymentDate,
    lastFinesPaymentDate,
    firstAssessmentDate,
    latestAssessmentDate,
    judgeNames,
  } = record.formInformation;

  if (ncicCheckDate) {
    transformedRecord.formInformation.ncicCheckDate =
      fieldToDate(ncicCheckDate);
  }

  if (lastRestitutionPaymentDate) {
    transformedRecord.formInformation.lastRestitutionPaymentDate = fieldToDate(
      lastRestitutionPaymentDate,
    );
  }

  if (lastFinesPaymentDate) {
    transformedRecord.formInformation.lastFinesPaymentDate =
      fieldToDate(lastFinesPaymentDate);
  }

  if (firstAssessmentDate) {
    transformedRecord.formInformation.firstAssessmentDate =
      fieldToDate(firstAssessmentDate);
  }

  if (latestAssessmentDate) {
    transformedRecord.formInformation.latestAssessmentDate =
      fieldToDate(latestAssessmentDate);
  }

  if (dateImposed) {
    transformedRecord.formInformation.dateImposed = dateImposed.map(
      (d: string) => fieldToDate(d),
    );
  }

  if (fullTermReleaseDates) {
    transformedRecord.formInformation.fullTermReleaseDates =
      fullTermReleaseDates.map((d: string) => fieldToDate(d));
  }

  if (judgeNames) {
    transformedRecord.formInformation.judgeNames = judgeNames.map(
      (blob: string) => {
        try {
          return Object.fromEntries(
            Object.entries(JSON.parse(blob)).map(([k, v]) => [camelCase(k), v]),
          );
        } catch (e) {
          return {};
        }
      },
    );
  }
  return transformedRecord;
};
