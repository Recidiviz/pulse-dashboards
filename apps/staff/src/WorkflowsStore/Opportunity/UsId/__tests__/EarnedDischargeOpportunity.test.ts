// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { DocumentData } from "firebase/firestore";
import { configure } from "mobx";
import tk from "timekeeper";

import { RootStore } from "../../../../RootStore";
import { Client } from "../../../Client";
import { DocumentSubscription } from "../../../subscriptions";
import {
  earnedDischargeAlmostEligibleSupervisionLength,
  EarnedDischargeEligibleClientRecord,
  EarnedDischargeReferralRecordFixture,
} from "../__fixtures__";
import { EarnedDischargeOpportunity } from "../EarnedDischargeOpportunity";

vi.mock("../../../subscriptions");

let opp: EarnedDischargeOpportunity;
let client: Client;
let root: RootStore;
let updatesSub: DocumentSubscription<any>;

function createTestUnit(
  clientRecord: typeof EarnedDischargeEligibleClientRecord,
  opportunityRecord: DocumentData,
) {
  root = new RootStore();
  root.workflowsRootStore.opportunityConfigurationStore.mockHydrated();
  vi.spyOn(root.workflowsStore, "opportunityTypes", "get").mockReturnValue([
    "earnedDischarge",
  ]);
  client = new Client(clientRecord, root);

  opp = new EarnedDischargeOpportunity(client, opportunityRecord);
}

beforeEach(() => {
  // this lets us spy on observables, e.g. computed getters
  configure({ safeDescriptors: false });
  tk.freeze(new Date(2022, 7, 1));
});

afterEach(() => {
  vi.resetAllMocks();
  tk.reset();
  configure({ safeDescriptors: true });
});

describe("fully eligible", () => {
  beforeEach(() => {
    createTestUnit(
      EarnedDischargeEligibleClientRecord,
      EarnedDischargeReferralRecordFixture,
    );

    updatesSub = opp.updatesSubscription;
    updatesSub.hydrationState = { status: "hydrated" };
  });

  test("requirements almost met", () => {
    expect(opp.almostEligible).toBeFalse();
    expect(opp.requirementsAlmostMet).toEqual([]);
  });

  test("requirements met", () => {
    expect(opp.requirementsMet).toMatchSnapshot();
  });
});

describe("almost eligible on probation at least a year", () => {
  beforeEach(() => {
    createTestUnit(
      EarnedDischargeEligibleClientRecord,
      earnedDischargeAlmostEligibleSupervisionLength,
    );

    updatesSub = opp.updatesSubscription;
    updatesSub.hydrationState = { status: "hydrated" };
  });

  test("requirements met", () => {
    expect(opp.requirementsMet).toMatchSnapshot();
  });

  test("requirements almost met", () => {
    expect(opp.almostEligible).toBeTrue();
    expect(opp.requirementsAlmostMet).toEqual([
      {
        key: "pastEarnedDischargeEligibleDate",
        text: "Needs 16 more months on supervision",
        tooltip:
          "If on probation, served minimum sentence according to the court; " +
          "if on parole for a nonviolent crime, served at least one year; if on parole for a sex/violent " +
          "offense, served at least one-third of remaining sentence; if on parole for a life sentence, served " +
          "at least five years on parole",
      },
    ]);
  });

  test("almostEligibleStatusMessage", () => {
    expect(opp.almostEligibleStatusMessage).toEqual(
      "Needs 16 more months on supervision",
    );
  });
});

describe("almost eligible days remaining on length of stay", () => {
  beforeEach(() => {
    tk.freeze(new Date("2022-01-23"));
    earnedDischargeAlmostEligibleSupervisionLength.ineligibleCriteria.onProbationAtLeastOneYear =
      { eligibleDate: "2022-01-24" };

    createTestUnit(
      EarnedDischargeEligibleClientRecord,
      earnedDischargeAlmostEligibleSupervisionLength,
    );

    updatesSub = opp.updatesSubscription;
    updatesSub.hydrationState = { status: "hydrated" };
  });

  test("requirements met", () => {
    expect(opp.requirementsMet).toMatchSnapshot();
  });

  test("requirements almost met", () => {
    expect(opp.almostEligible).toBeTrue();
    expect(opp.requirementsAlmostMet).toEqual([
      {
        key: "pastEarnedDischargeEligibleDate",
        text: "Needs 1 more day on supervision",
        tooltip:
          "If on probation, served minimum sentence according to the court; " +
          "if on parole for a nonviolent crime, served at least one year; if on parole for a sex/violent " +
          "offense, served at least one-third of remaining sentence; if on parole for a life sentence, served " +
          "at least five years on parole",
      },
    ]);
  });

  test("almostEligibleStatusMessage", () => {
    expect(opp.almostEligibleStatusMessage).toEqual(
      "Needs 1 more day on supervision",
    );
  });
});
