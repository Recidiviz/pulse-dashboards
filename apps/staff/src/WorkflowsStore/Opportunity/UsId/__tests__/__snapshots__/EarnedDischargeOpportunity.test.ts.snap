// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`almost eligible days remaining on length of stay > requirements met 1`] = `
[
  {
    "key": "usIdLsirLevelLowModerateForXDays",
    "text": "Currently moderate risk with no increase in risk level in past 360 days",
    "tooltip": "Assessed at low risk level on LSI-R with no risk increase in past 90 days or moderate risk level on LSI-R with no risk increase in past 360 days",
  },
  {
    "key": "negativeDaWithin90Days",
    "text": "Negative UA within past 90 days",
    "tooltip": "Negative UA within past 90 days, unless the client lacks a history of drug/alcohol abuse or has been supervised at low risk for more than one year",
  },
  {
    "key": "noFelonyWithin24Months",
    "text": "No felony convictions in past 24 months",
    "tooltip": "Has not committed a felony while on probation or parole in past 24 months",
  },
]
`;

exports[`almost eligible on probation at least a year > requirements met 1`] = `
[
  {
    "key": "usIdLsirLevelLowModerateForXDays",
    "text": "Currently moderate risk with no increase in risk level in past 360 days",
    "tooltip": "Assessed at low risk level on LSI-R with no risk increase in past 90 days or moderate risk level on LSI-R with no risk increase in past 360 days",
  },
  {
    "key": "negativeDaWithin90Days",
    "text": "Negative UA within past 90 days",
    "tooltip": "Negative UA within past 90 days, unless the client lacks a history of drug/alcohol abuse or has been supervised at low risk for more than one year",
  },
  {
    "key": "noFelonyWithin24Months",
    "text": "No felony convictions in past 24 months",
    "tooltip": "Has not committed a felony while on probation or parole in past 24 months",
  },
]
`;

exports[`fully eligible > requirements met 1`] = `
[
  {
    "key": "usIdLsirLevelLowModerateForXDays",
    "text": "Currently moderate risk with no increase in risk level in past 360 days",
    "tooltip": "Assessed at low risk level on LSI-R with no risk increase in past 90 days or moderate risk level on LSI-R with no risk increase in past 360 days",
  },
  {
    "key": "negativeDaWithin90Days",
    "text": "Negative UA within past 90 days",
    "tooltip": "Negative UA within past 90 days, unless the client lacks a history of drug/alcohol abuse or has been supervised at low risk for more than one year",
  },
  {
    "key": "noFelonyWithin24Months",
    "text": "No felony convictions in past 24 months",
    "tooltip": "Has not committed a felony while on probation or parole in past 24 months",
  },
  {
    "key": "pastEarnedDischargeEligibleDate",
    "text": "Served minimum sentence requirements: has served 2 years, 7 months",
    "tooltip": "If on probation, served minimum sentence according to the court; if on parole for a nonviolent crime, served at least one year; if on parole for a sex/violent offense, served at least one-third of remaining sentence; if on parole for a life sentence, served at least five years on parole",
  },
]
`;
